Add more tests to vesync

iprak opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This request added more tests to get 100% coverage for humidifier.

There was one function `unique_info` which was appearing uncovered but was not used anywhere, it has been removed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iprak committed:
Added discovery test

iprak committed:
Removed unique_info

This was not being use anywhere.

iprak committed:
Added a test for base_unique_id

The device data for humidifier "Humidifier 200s" has subDeviceNo defined
which would get appended to unique_id.

iprak committed:
Split available_modes and added tests

joostlek commented on homeassistant/components/vesync/fan.py:
```diff
@@ -151,11 +151,6 @@ def preset_mode(self) -> str | None:
-        """Return the ID of this fan."""
-        return self.smartfan.uuid
-
     @property
```
why?

iprak replied:
It was not used anywhere.

joostlek replied:
Oh I misread as `unique_id`

joostlek commented on homeassistant/components/vesync/humidifier.py:
```diff
@@ -96,6 +96,15 @@ def _get_vs_mode(ha_mode: str) -> str | None:
     return HA_TO_VS_MODE_MAP.get(ha_mode)
 
 
+def _get_available_modes(mist_modes: list[str]) -> list[str]:
+    """Return the available mist modes."""
+    return [
+        ha_mode
+        for ha_mode in (_get_ha_mode(vs_mode) for vs_mode in mist_modes)
+        if ha_mode
+    ]
```
why?

iprak replied:
I want to test the function separately from device.

joostlek replied:
We should still mock a device, as we don't want to mock internals

iprak replied:
The condition cannot be replicated by mocking a device.  `available_modes` is calculated based on `device.mist_modes` which is a hard coded in [pyvesync](https://github.com/webdjoe/pyvesync/blob/535d5324727546037f9d37bce4dbed4358e501a5/src/pyvesync/vesyncfan.py#L22). I think mocking something internal to pyvesync would be worse.

iprak replied:
Rewrote the test with mocked humidifier device.

joostlek commented on tests/components/vesync/test_entity.py:
```diff
@@ -0,0 +1,23 @@
+NoException = nullcontext()
+
+
+async def test_base_unique_id(
+    hass: HomeAssistant,
+    humidifier_config_entry: MockConfigEntry,
+    entity_registry: er.EntityRegistry,
+) -> None:
+    """Test that unique_id is based on subDeviceNo."""
+    # vesync-device.json defines subDeviceNo for 200s-humidifier as 4321.
+    entity = entity_registry.async_get(ENTITY_HUMIDIFIER)
+    assert entity.unique_id.endswith("4321")
```
this is a humidifier test, not an enttiy one

iprak replied:
I have moved it to test_humidifier

joostlek commented on tests/components/vesync/test_humidifier.py:
```diff
@@ -55,6 +62,33 @@ async def test_humidifier_state(
     assert state.attributes.get(ATTR_HUMIDITY) == 40
 
 
+async def test_invalid_available_modes() -> None:
+    """Test invalid available mode mapping."""
+
+    with patch(
+        "homeassistant.components.vesync.humidifier._LOGGER.warning"
+    ) as mocked_warning:
+        _get_available_modes(["invalid_mode"])
+        mocked_warning.assert_called_once()
+
+
+async def test_valid_available_modes() -> None:
+    """Test valid available mode mapping."""
+
+    with patch(
+        "homeassistant.components.vesync.humidifier._LOGGER.warning"
+    ) as mocked_warning:
+        _get_available_modes(
+            [
+                VS_HUMIDIFIER_MODE_AUTO,
+                VS_HUMIDIFIER_MODE_HUMIDITY,
+                VS_HUMIDIFIER_MODE_MANUAL,
+                VS_HUMIDIFIER_MODE_SLEEP,
+            ]
+        )
+        mocked_warning.assert_not_called()
```
don't import the function. You can check the log output via caplog

iprak committed:
Added test_base_unique_id to test_humidifier

iprak committed:
Using caplog for testing logging

iprak committed:
Verify that device mist mode is valid

iprak committed:
Rebased and removed previous tests

Removed tests which were using pyvesync.vesyncfan.model_features.

iprak committed:
Return None as the mode if device.mode is None

VeSyncBaseDevice sets mode to details.get('mode') which happens
to be None and ends up logging incorrect warning.

iprak committed:
Added tests for device mist mode

This uses a mocked VeSyncHumid200300S device with mist_modes set to
test values.
